transport/ice/IceSession.cpp

```diff
@@ -1091,31 +1270,172 @@ void IceSession::stateCheck(const uint64_t now)
+        {
+            logger::debug("freezePendingProbes - remove failed candidate: %s %s %s - %s",
+                _logId.c_str(),
+                toString(candidatePair->localEndpoint.endpoint->getTransportType()).c_str(),
+                toString(candidatePair->remoteCandidate.type).c_str(),
+                candidatePair->localCandidate.address.toString().c_str(),
+                candidatePair->remoteCandidate.address.toString().c_str());
+
+            onCandidatePairRemoved(candidatePair);
+            std::iter_swap(it, --itEnd);
+            continue;
+        }
+
+        addCandidateToListIfNotPresent(_remoteCandidates, candidatePair->remoteCandidate);
+        ++it;
+    }
+
+    _candidatePairs.erase(itEnd, _candidatePairs.end());
+}
+
+void IceSession::removeNonReadableRemoteCandidates(uint64_t now)
+{
+    // If we are in the limit, we will be more aggressive on removing candidates
+    // by removing HOST and SRFLX candidates that didn't work or are inactive for long period.
+    // Wen the list is not full, we will only remove PRFLX
+    const bool aggressiveRemove = _remoteCandidates.size() >= _config.maxCandidateCount;
+    _remoteCandidates.clear();
+
+    auto itEnd = _candidatePairs.end();
+
+    for (auto it = _candidatePairs.begin(); it != itEnd;)
+    {
+        auto* candidatePair = it->get();
+
+        // Let's keep the nominated regardless the current state
+        if (candidatePair->nominated)
+        {
+            addCandidateToListIfNotPresent(_remoteCandidates, candidatePair->remoteCandidate);
+            ++it;
+            continue;
+        }
+
+        // Keep HOST and SRFLX if we are not in aggressive removing.
+        if (!aggressiveRemove && candidatePair->remoteCandidate.type != IceCandidate::Type::PRFLX)
+        {
+            addCandidateToListIfNotPresent(_remoteCandidates, candidatePair->remoteCandidate);
+            ++it;
+            continue;
+        }
+
+        if (candidatePair->isReadable(now))
+        {
+            addCandidateToListIfNotPresent(_remoteCandidates, candidatePair->remoteCandidate);
+            ++it;
+            continue;
+        }
+
+        logger::info("Remove unreadable candidate pair: %s %s %s - %s",
+            _logId.c_str(),
+            toString(candidatePair->localEndpoint.endpoint->getTransportType()).c_str(),
+            toString(candidatePair->remoteCandidate.type).c_str(),
+            candidatePair->localCandidate.address.toString().c_str(),
+            candidatePair->remoteCandidate.address.toString().c_str());
+
+        candidatePair->state = ProbeState::Failed;
+        std::iter_swap(it, --itEnd);
+        onCandidatePairRemoved(candidatePair);
+    }
+
+    _candidatePairs.erase(itEnd, _candidatePairs.end());
+
+    // If after the first scan the list it still bigger or equal to _config.maxCandidateCount,
+    // then we will be even more aggressive on removing candidates
+    if (_remoteCandidates.size() <= _config.maxCandidateCount)
```

Comment:
If after first round, we still have a full candidate list, then we going to try to release some of them. I am not truly confident that we should do this but I am not sure if rejecting a new candidate is actually a better option.

Then the algorithm fist sort the candidatesPair. First the dominated, then the ones in a readable state ordered by priority (where we receive packet in the latest 45 seconds. Then the other ones ordered by priority.

Then we going to read them to `_remoteCandidates ` until a point that the list is not full and remove the lower priority ones that didn't fit